### DISCUSSION TITLE
Performing unattended install

### DISCUSSION THREAD

**Discussion by ga2k** [abraunegg/onedrive#1793] [Q&A] [answered]
Hi, I've finished writing a mammoth bash script that fully automates the creation of VMWare VM's from command line parameters. It goes from zero to a fully functioning OS complete with your onedrive client having done its initial sync. [ And wasn't it a job to get the dmd installer to download the source and compile onedrive without pause... ]

And the best bit is it is completely hands off. Until it comes to the part (lots of time into the run) when the onedrive client has to be authenticated and the token retrieved. So the user has to sit and watch for 20 odd minutes, then copy the URL into a browser, authenticate, and get the response, then copy that and paste it back in.

Is there a way to automate this step, too? Either at the point where it is required, or before the whole process begins? I tried to send the url using lynx but I didn't get far with that.

Once this part is fixed, my job is done!

What is in that URI? Can that be fabricated before we begin?

Any ideas?

**Comment by abraunegg** [comment, marked as answer]
@ga2k 
> Hi, I've finished writing a mammoth bash script that fully automates the creation of VMWare VM's from command line parameters. It goes from zero to a fully functioning OS complete with your onedrive client having done its initial sync. [ And wasn't it a job to get the dmd installer to download the source and compile onedrive without pause... ]

Sounds like quite a process.

> And the best bit is it is completely hands off. Until it comes to the part (lots of time into the run) when the onedrive client has to be authenticated and the token retrieved. So the user has to sit and watch for 20 odd minutes, then copy the URL into a browser, authenticate, and get the response, then copy that and paste it back in.
> 
> Is there a way to automate this step, too? Either at the point where it is required, or before the whole process begins? I tried to send the url using lynx but I didn't get far with that.

The authentication URL is fairly static, unless you are configuring a specific Azure authentication domain and/or changing the application ID. Outside of that, the URL is not going to change, so it could be referenced and used / accessed earlier than when prompted.

> What is in that URI? Can that be fabricated before we begin?

You can get the response URI before hand, then utilise the auth files option to pass in the response:
```
--auth-files ARG
      Perform authorization via two files passed in as ARG in the format `authUrl:responseUrl`
      The authorization URL is written to the `authUrl`, then onedrive waits for the file `responseUrl`
      to be present, and reads the response from that file.
```
This is how the authentication is done in some Docker compose environments

**Reply by MrJohanssons** [reply]
Hi,

We are tying to implement this but it's not working. It seems like the response URL is unique. How can we get it before hand?


```
onedrive --auth-files 'authUrl:responseUrl'

Configuring Global Azure AD Endpoints
ERROR: Microsoft OneDrive API returned an error with the following message:
  Error Message:    HTTP request returned status code 400 (Bad Request)
  Error Reason:     AADSTS54005: OAuth2 Authorization code was already redeemed, please retry with a new valid code or use an existing refresh token.
Trace ID: < redacted >
Correlation ID: < redacted>
Timestamp: 2023-01-24 10:08:37Z

--synchronize or --monitor switches missing from your command line input. Please add one (not both) of these switches to your command line or use 'onedrive --help' for further assistance.

No OneDrive sync will be performed without one of these two arguments being present.
```

**Reply by abraunegg** [reply]
@MrJohanssons 
> `onedrive --auth-files 'authUrl:responseUrl'`

The 'authUrl' is a file locally that will contain the URL that needs to be accessed to obtain the authentication response

The 'responseUrl' is a file locally that will contain the response URI to the the above URL

`onedrive --auth-files '/path/to/auth_url:/path/to/auth_response'`

The application will wait for the '/path/to/auth_response' to be present, and, when it is present it needs to contain the correct response provided by the Auth URL.

The error you are getting indicates to me that you are not using the correct Auth URL with the right token and/or the correct response in the response file - essentially trying to re-use an auth response which has already been used.

**Reply by MrJohanssons** [reply]
> The error you are getting indicates to me that you are not using the correct Auth URL with the right token and/or the correct response in the response file - essentially trying to re-use an auth response which has already been used.

Yes, I probably misunderstood the instructions. I got the auth url from onedrive cli output and response url from the browser. So that explains why it says the code has already been redeemed.

So what I don't understand is how do I get the response URL? The auth url should be static right? But the response url changes each time?

**Reply by abraunegg** [reply]
@MrJohanssons 
> Yes, I probably misunderstood the instructions. I got the auth url from onedrive cli output and response url from the browser.

That is the correct process.

> So what I don't understand is how do I get the response URL? The auth url should be static right? But the response url changes each time?

When you use `onedrive --auth-files '/path/to/auth_url:/path/to/auth_response'` the client will wait for the file to be present. When it is present it needs to contain the output that was presented to you when you provided the Auth URL to the browser and provided the credentials.

Essentially, in a different terminal window | SSH session:
```
echo "https://login.microsoftonline.com/common/oauth2/nativeclient?code=<redacted>" > /path/to/auth_response
```

This will write your auth response as provided by the browser to the file you are specifying when using those client options.

How you 'get' that auth URL ..... well you need to use a web browser as that is the Microsoft process.

**Reply by MrJohanssons** [reply]
Maybe it's possible to get these URLs with _**cURL**_? Using a browser will be hard to script. @ga2k how did you solve this? Maybe you can share some snippets from your automation script?

I guess the biggest problem will be the authentication prior to receiving the response URL.

**Comment by ga2k** [comment]
Thanks for that. I was sure I has seen the command line option for that, but I was blind! It works an absolute treat.

Cheers,
Geoffrey.


On 13 Jan 2022, at 5:51 pm, abraunegg ***@***.******@***.***>> wrote:



@ga2k<https://github.com/ga2k>

Hi, I've finished writing a mammoth bash script that fully automates the creation of VMWare VM's from command line parameters. It goes from zero to a fully functioning OS complete with your onedrive client having done its initial sync. [ And wasn't it a job to get the dmd installer to download the source and compile onedrive without pause... ]

Sounds like quite a process.

And the best bit is it is completely hands off. Until it comes to the part (lots of time into the run) when the onedrive client has to be authenticated and the token retrieved. So the user has to sit and watch for 20 odd minutes, then copy the URL into a browser, authenticate, and get the response, then copy that and paste it back in.

Is there a way to automate this step, too? Either at the point where it is required, or before the whole process begins? I tried to send the url using lynx but I didn't get far with that.

The authentication URL is fairly static, unless you are configuring a specific Azure authentication domain and/or changing the application ID. Outside of that, the URL is not going to change, so it could be referenced and used / accessed earlier than when prompted.

What is in that URI? Can that be fabricated before we begin?

You can get the response URI before hand, then utilise the auth files option to pass in the response:

--auth-files ARG
      Perform authorization via two files passed in as ARG in the format `authUrl:responseUrl`
      The authorization URL is written to the `authUrl`, then onedrive waits for the file `responseUrl`
      to be present, and reads the response from that file.


This is how the authentication is done in some Docker compose environments

—
Reply to this email directly, view it on GitHub<https://github.com/abraunegg/onedrive/discussions/1793#discussioncomment-1960581>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AGDW5FPNMQKCUZ5QEU3IFYDUVZ77JANCNFSM5L23EK3A>.
Triage notifications on the go with GitHub Mobile for iOS<https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675> or Android<https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
